patches/server/0722-Rate-options-and-timings-for-sensors-and-behaviors.patch

```diff
@@ -0,0 +1,190 @@
++        return Timings.ofSafe("Behavior - " + type);
++    }
++
++    public static Timing getSensorTimings(String type) {
++        return Timings.ofSafe("Sensor - " + type);
++    }
++
+     /**
+      * Get a named timer for the specified tile entity type to track type specific timings.
+      * @param entity
+diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+index 26e18a08a7f0bd704ff3055ce3a7814191450c85..2a3be23f52de7a960e0e323dd507579422b5b30d 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -3,6 +3,7 @@ package com.destroystokyo.paper;
+ import java.util.Arrays;
+ import java.util.HashMap;
+ import java.util.List;
++import java.util.Locale;
+ import java.util.Map;
+ import java.util.stream.Collectors;
+ import net.minecraft.world.Difficulty;
+@@ -11,6 +12,7 @@ import net.minecraft.world.entity.monster.Vindicator;
+ import net.minecraft.world.entity.monster.Zombie;
+ import com.destroystokyo.paper.antixray.ChunkPacketBlockControllerAntiXray.EngineMode;
+ import org.bukkit.Bukkit;
++import org.bukkit.configuration.ConfigurationSection;
+ import org.bukkit.configuration.file.YamlConfiguration;
+ import org.spigotmc.SpigotWorldConfig;
+ 
+@@ -820,5 +822,33 @@ public class PaperWorldConfig {
+     private void fixInvulnerableEndCrystalExploit() {
+         fixInvulnerableEndCrystalExploit = getBoolean("unsupported-settings.fix-invulnerable-end-crystal-exploit", fixInvulnerableEndCrystalExploit);
+     }
++
++    private Map<String, Integer> tickRates = new HashMap<>();
++    private void tickRates() {
++        config.addDefault("world-settings.default.tick-rates.sensor.villager.secondaryplaces", 40);
++        config.addDefault("world-settings.default.tick-rates.behavior.villager.positionvalidate", 20);
++        tickRates = new HashMap<>();
++        log("Tick rates:");
++        ConfigurationSection tickRatesSection = config.getConfigurationSection("world-settings." + worldName + ".tick-rates");
++        if (tickRatesSection == null) {
++            tickRatesSection = config.getConfigurationSection("world-settings.default.tick-rates");
++        }
++        if (tickRatesSection != null) {
++            for (String key : tickRatesSection.getKeys(true)) {
++                if (tickRatesSection.isInt(key)) {
++                    int tickRate = tickRatesSection.getInt(key);
++                    tickRates.put(key.toLowerCase(Locale.ROOT), tickRate);
++                    log("  " + key + ": " + tickRate);
++                }
++            }
++        }
++        if (tickRates.isEmpty()) {
++            log("  None configured");
++        }
++    }
++    public int getTickRate(String type, String typeName, String entityType, int def) {
++        int rate = tickRates.getOrDefault(type + "." + entityType + "." + typeName, -1);
```

Comment:
This likely will cause a lot of memory churning due to the string concatenation, it would be worth caching.

Reply:
Not sure what would be the best way to cache that. I guess simply nesting maps/tables on config load could be best?

Reply:
Given that (iirc) reloading the config isn't officially supported for everything, you could always just store the tick rate once in the Behavior/Sensor classes.

Reply:
Caching it in the classes is not possible as Behaviors and Sensors only have one global instance that is shared between all entity types and accessed via their types. (See `net.minecraft.world.entity.ai.sensing.SensorType` for example)